{% extends "base.html" %}

{% block content %}

    <div class="container">

        <div class="page-header">
            <h1>About HmtVar</h1>
        </div>
        <p class="lead">HmtVar is a free resource hosting variability and pathogenicity data about human mitochondrial variants; this information is also integrated with data coming from several online databases and tools focused on the mitochondrial genome.</p>

        <div class="container">
        <div class="row">
            <h2>Database content</h2>
            <p>HmtVar contains variants data coming primarily from complete human mitochondrial genomes stored in the <a href="https://www.hmtdb.uniba.it" title="HmtDB" target="_blank">HmtDB</a> online resource.</p>
            <p>HmtVar variants are gathered from variantions found in complete human mitochondrial genomes downloaded from GenBank in HmtDB and derived from individuals with apparently healthy phenotype and individuals with disease phenotypes (annotated in HmtDB as "healthy" and "pathologic" respectively). <small>Caveat: genomes available in HmtDB also include coding-region-only sequences, whose variants are not included in HmtVar, since these genomes are not complete.</small> </p>
            <p>Variants are determined based on the <a href="https://www.ncbi.nlm.nih.gov/nuccore/251831106" title="rCRS" target="_blank">rCRS</a> mitochondrial reference sequence.</p>
            <p>HmtVar offers a pathogenicity prediction for both non synonymous and tRNA variants; these predictions are calculated using variants nucleotide variability and specific <a href="{{ url_for('site.pathogenicity') }}" title="Pathogenicity criteria">pathogenicity criteria</a>, as detailed in <a href="https://link.springer.com/article/10.1007%2Fs00439-015-1615-9" title="Santorsola et al" target="_blank">Santorsola et al</a> for non-synonymous variants and in <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-016-1193-4" title="Diroma et el" target="_blank">Diroma et al</a> for tRNA variants.</p>
            <p>tRNA variants attributes are manually curated and further information can be found on the <a href="{{ url_for('site.trna_models') }}" title="tRNA Models">tRNA models</a> page. </p>

            <h3>Database structure</h3>

            <p>Each variant in HmtVar is associated to the following attributes, which allow an exhaustive data retrieval through both the
                <a href="{{ url_for('site.query') }}" title="Query">Query</a> page or the <a href="{{ url_for('site.apis') }}" title="API">API</a> functionalities.</p>

            <div class="container">
            <div class="row">

                <div class="col-sm-12">
                    <h4>Main Information</h4>
                    <p>Allows to query the basic information regarding variants; these data include:</p>
                    {% include "db_structure/main_table.html" %}
                </div>

            </div>

            <div class="row">

                <div class="col-sm-12">
                    <h4>Variability</h4>
                    <p>Allows to query the variability information regarding variants; these data include:</p>
                    {% include "db_structure/variab_table.html" %}
                </div>

            </div>
            </div>

            <h2>Variant Card</h2>

            <p>After performing a query, detailed information about a specific variant is shown in a Variant Card, which gathers all the data available using different tabs. These tabs are: </p>

            <div class="container">

            <ul class="nav nav-tabs">
                <li class="active"><a data-toggle="tab" href="#main_tab">Main Info</a></li>
                <li><a data-toggle="tab" href="#var_tab">Variability</a></li>
                <li><a data-toggle="tab" href="#predict_tab">Pathogenicity Predictions</a></li>
                <li><a data-toggle="tab" href="#external_tab">External Resources</a></li>
                <li><a data-toggle="tab" href="#downloads_tab">Download Data</a></li>
            </ul>

            <div class="tab-content">

                <div id="main_tab" class="tab-pane fade in active">
                    <br><h3>Main Info</h3>
                    <p>Contains variant’s basic information such as its location, the consequent amino acidic change for non-synonymous variants or specific structural details for tRNA variants, haplogroups and macro-haplogroups code if the variant is associated to a specific haplogroup according to <a href="http://phylotree.org/index.htm" title="Phylotree" target="_blank">Phylotree build 17</a>, as well as HmtVar pathogenicity prediction.</p>
                </div>

                <div id="var_tab" class="tab-pane fade">
                    <br><h3>Variability</h3>
                    <p>Shows nucleotide and amino-acidic variability values, for both healthy and diseased genomes; in the same tab, healthy, diseased and continent-specific allele frequencies are also reported.</p>
                </div>

                <div id="predict_tab" class="tab-pane fade">
                    <br><h3>Pathogenicity Predictions</h3>
                    <p>Shows a more detailed view of pathogenicity annotated in HmtVar inclusive of the global disease score, as well as pathogenicity scores calculated by the predictors used to estimate the DS value.</p>
                </div>

                <div id="external_tab" class="tab-pane fade">
                    <br><h3>External Resources</h3>
                    <p>Shows a set of additional variant information coming from online resources focused on diseases (Mitomap, ClinVar, OMIM), population studies (dbSNP and 1000Genomes) and structural information (Mamit-tRNA). Links to the original sources of information are always provided for consistency. </p>
                </div>

                <div id="downloads_tab" class="tab-pane fade">
                    <br><h3>Download Data</h3>
                    <p>Allows users to download each variant’s data for offline use. Data are provided in a JSON-formatted file, which contains all the information available for the variant selected.</p>
                </div>

            </div>

            </div>

            <h2>Citing HmtVar</h2>

            <p>HmtVar can be cited using the latest publication, available on <a href="https://doi.org/10.1093/nar/gky1024" title="HmtVar NAR" target="_blank">Nucleic Acids Research</a>: </p>

            <div class="panel panel-success">
                <div class="panel-body">
                    <p>Preste R, Vitale O, Clima R, Gasparre G, Attimonelli M</p>
                    <p><b>HmtVar: a new resource for human mitochondrial variations and pathogenicity data.</b></p>
                    <p><a href="https://doi.org/10.1093/nar/gky1024" title="HmtVar NAR" target="_blank"><i>Nucleic Acids Res. 2018 Oct 29.</i></a></p>
                </div>
            </div>
            <p>Code for HmtVar can be found on the official <a href="https://github.com/robertopreste/HmtVar" title="HmtVar GitHub" target="_blank">GitHub repository</a>, where a subset of the actual data is available to test the database and play with it.</p>
            <br><br>
        </div>
        </div>

    </div>

    {{ super() }}
{% endblock %}
